libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2950,6 +2945,11 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 				"\"chainid\" is not supported by the VM version."
 			);
 	}
+	else if (m_currentContract || m_currentFreeFunction)
+		if (auto typeType = dynamic_cast<TypeType const*>(exprType))
+			if (auto contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+				if (&contractType->contractDefinition() != m_currentContract)
+					currentContractDependencies()->contractDependencies.emplace(&contractType->contractDefinition(), &_memberAccess);
```

Comment:
That's true. I think I added the `selector` check due to a request from @chriseth. Chris, maybe you can clear up the desired behavior here? :)

Reply:
I don't really understand the question here. There is no contract creation involved in the example, so there should be no cycle, not even an edge.

Reply:
But it's accessing attributes of a different contract, that is a dependency, right?

Reply:
> I don't really understand the question here. There is no contract creation involved in the example, so there should be no cycle, not even an edge.

I agree there should be no cycle. I think that the piece of code I commented on would result in it being interpreted as a cycle so we have to remove or at least adjust it.

> But it's accessing attributes of a different contract, that is a dependency, right?

`selector` has special rules in that regard. For example you can access selectors of functions from a contract you do not inherit from. It's allowed because you don't have to compile that whole contract before yours to calculate the selector.

`C.f.selector` does result in `f` being added to internal dispatch but I don't think it counts as a dependency because you then just compile this function as a part of your contract. And you could make `s` in my example `constant` which would not even add it to the dispatch.

This case came up recently in #11015 and #11014.